Feature/sc 1585/anomaly generation

ssirnio opened the pull request:
## Jira Story
[SC-1585](https://epcpower.atlassian.net/browse/SC-1585)

## About
Adds anomaly generation for anomaly supervision service. See more details on m-tcu PR.

## Associated Pull Requests

* [ ] _https://github.com/epcpower/m-tcu/pull/20
* [ ] _

## Reproduction Steps

## Validation


[SC-1585]: https://epcpower.atlassian.net/browse/SC-1585?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

ssirnio committed:
feat(SC-1585): Add anomaly editor for PM

ssirnio committed:
feat(SC-1585): Add C-code and spreadsheet export functions for anomalies

ssirnio committed:
chore(SC-1585): Run black formatter

ssirnio committed:
fix(SC-1585): Adjust tests to check for anomalies

gordon-epc commented on src/epcpm/anomaliestoxlsx.py:
```diff
@@ -0,0 +1,275 @@
+    desc="Description",
+)
+
+"""
+Exports anomaly information to a spreadsheet file.
+
+Args:
+    path:             Path to output XLSX file.
+    anomaly_model:    Anomaly data model.
+    parameters_model: Parameter data model.
+    column_filter:    Optional filter to remove selected output columns.
+    skip_output:      Optional boolean flag to disable output.
+
+Returns:
+    None
+"""
+
+
+def export(
+    path,
+    anomaly_model,
+    parameters_model,
+    column_filter=None,
+    skip_output=False,
+):
```
The style is to put docstrings below the `def`. I only put a review comment on this one, but there are others throughout this PR.

```suggestion

def export(
    path,
    anomaly_model,
    parameters_model,
    column_filter=None,
    skip_output=False,
):
    """
    Exports anomaly information to a spreadsheet file.

    Args:
        path:             Path to output XLSX file.
        anomaly_model:    Anomaly data model.
        parameters_model: Parameter data model.
        column_filter:    Optional filter to remove selected output columns.
        skip_output:      Optional boolean flag to disable output.

    Returns:
        None
    """

```

gordon-epc commented on src/epcpm/anomaliestoxlsx.py:
```diff
@@ -0,0 +1,275 @@
+                    row.as_filtered_tuple(self.column_filter),
+                )
+
+        # Adjust column widths in the worksheet in regards of text lenght
```
```suggestion
        # Adjust column widths in the worksheet in regards of text length
```

gordon-epc commented on src/epcpm/project.py:
```diff
@@ -217,6 +284,46 @@ def update_enumeration_roots(self):
             )
         self.parameters.list_selection_roots["access level"] = access_level_root
 
+        # Save location of anomaliy code enumeration
```
```suggestion
        # Save location of anomaly code enumeration
```

gordon-epc commented on src/epcpm/tests/project/anomalies.json:
It would be good to have sample anomalies to unit test.

gordon-epc replied:
I still recommend a basic unit test case in `src/epcpm/tests/project/anomalies.json`.

ssirnio replied:
Some simple tests added for model loading and spreadsheet generation

gordon-epc replied:
Great!

gordon-epc reviewed (changes requested):
My comments are superficial at this point. I haven't had a chance to run the code to see what happens. I will run the code next chance I get.

I know that old PM code is lacking type hints, but for new code, I made an effort to use type hints. Since the plan is to use this code for a future PM, whatever that may look like, we should use type hints.

vvainola commented:
Does the target branch have to be m-tcu instead of develop? Is this change incompatible with gridtied unless empty anomalies.json is added there?

ssirnio commented:
> Does the target branch have to be m-tcu instead of develop? Is this change incompatible with gridtied unless empty anomalies.json is added there?

Some of the tests inside pm repo, which use an example project similar to gridtied, had issues unless the `anomalies.json` was added. I didn't check what would happen with gridtied, but my assumption is that similar issues would rise. But I agree it would be better not to diverge pm further between different projects.

ssirnio committed:
fix(SC-1585): typo

Co-authored-by: Gordon Schmidt <77705928+gordon-epc@users.noreply.github.com>

ssirnio committed:
fix(SC-1585): typo

Co-authored-by: Gordon Schmidt <77705928+gordon-epc@users.noreply.github.com>

ssirnio committed:
fix(SC-1585): better human-readable names

ssirnio committed:
chore(SC-1585): fix docstrings

ssirnio committed:
chore(SC-1585): add type hints

gordon-epc commented on src/epcpm/anomaliestoc.py:
```diff
@@ -0,0 +1,104 @@
+
+@builders(epcpm.anomalymodel.AnomalyTable)
+@attr.s
+class AnomalyTable:
```
Would it be better for the vocabulary to be "AnomalyGroup" instead of "AnomalyTable"?

ssirnio replied:
On firmware repos m-tcu and rm-dc, they are consistently named "Tables", so I think it would be better to keep that name to avoid confusion.

gordon-epc commented on src/epcpm/anomaliestoc.py:
```diff
@@ -0,0 +1,104 @@
+            "name": self.wrapped.name,
+            "code": self.wrapped.code,
+            "trigger_type": trig_type,
+            "response_level_I": resp_level_I,
+            "response_level_A": resp_level_A,
```
Is there much gained by shortening "response_level_I" and "response_level_A"? Why not just "response_level_inactive" and "response_level_active"?

ssirnio replied:
I'll remove the shortening, I agree it's confusing.

gordon-epc commented on src/epcpm/anomaliestoxlsx.py:
```diff
@@ -0,0 +1,277 @@
+
+
+field_names = Fields(
+    name="Name",
+    group="Group",
```
In the spreadsheet, does it make more sense to have the Group be the first column and Name be second?

ssirnio replied:
Done

gordon-epc commented on src/epcpm/anomalymodel.py:
```diff
@@ -0,0 +1,156 @@
+        )
+
+
+@graham.schemify(tag="anomaly_table")
```
Again, `anomaly_group `seems like better vocabulary.

gordon-epc commented on src/epcpm/importexport.py:
```diff
@@ -311,6 +314,19 @@ def full_export(
+        path=paths.anomalies_spreadsheet,
+        anomaly_model=project.models.anomalies,
+        parameters_model=project.models.parameters,
+        skip_output=skip_output,
```
Just so you know, `skip_output` is true when the SunSpec output is skipped. This may or may not be the behavior you want?

ssirnio replied:
Ok, if it is sunspec-specific flag I won't utilize it in anomaly generation

gordon-epc commented on src/epcpm/mainwindow.ui:
```diff
@@ -30,13 +30,13 @@
       <widget class="QTabWidget" name="tabWidget">
        <property name="currentIndex">
-        <number>0</number>
+        <number>5</number>
```
Leave the CAN tab as the selected tab at application startup.
```suggestion
        <number>0</number>
```

gordon-epc commented on src/epcpm/mainwindow.ui:
```diff
@@ -121,7 +121,24 @@
+            <widget class="QTreeView" name="anomalies_view"/>
+           </item>
+           <item row="0" column="0">
+            <widget class="SearchBox" name="anomalies_search_box_2"/>
```
The anomalies search box does not work. I'm unsure which column it should be searching on?
```suggestion
            <widget class="SearchBox" name="anomalies_search_box"/>
```

ssirnio replied:
Fixed, will search on name column

gordon-epc commented on src/epcpm/mainwindow.ui:
```diff
@@ -151,8 +168,8 @@
      <y>0</y>
-     <width>1200</width>
-     <height>20</height>
+     <width>1516</width>
+     <height>21</height>
```
Unless there is a reason to change the size, I'd leave these alone.

ssirnio replied:
These changes were added automatically by the Qt designer tool. Reverted.

gordon-epc commented on src/epcpm/mainwindow.ui:
```diff
@@ -6,8 +6,8 @@
     <y>0</y>
-    <width>1200</width>
-    <height>900</height>
+    <width>1516</width>
+    <height>983</height>
```
Unless there is a reason to change the size, I'd leave these alone.

ssirnio replied:
These changes were added automatically by the Qt designer tool. Reverted.

ssirnio committed:
fix(SC-1585): remove abbreviation

ssirnio committed:
fix(SC-1585): anomaly table name is the first column in spreadsheet

ssirnio committed:
fix(SC-1585): revert unnecessary pm ui changes

ssirnio committed:
fix(SC-1585): CAN tab should be default

Co-authored-by: Gordon Schmidt <77705928+gordon-epc@users.noreply.github.com>

ssirnio committed:
fix(SC-1585): search box name

Co-authored-by: Gordon Schmidt <77705928+gordon-epc@users.noreply.github.com>

ssirnio committed:
fix(SC-1585): search box

ssirnio committed:
fix(SC-1585): always output anomalies spreadsheet

ssirnio committed:
feat(SC-1585): Anomaly sources can be added under anomalies

Also a text field for explaining the meaning of "detail" of
an anomaly was added.

ssirnio committed:
feat(SC-1585): Add tests for anomaly generation
